### ISSUE TITLE
SyntaxError: 'return' with argument inside generator

### ISSUE THREAD

**Issue by GForceSquared** [pootle/pisteppers#1]
Hi, I was using your code and I tried to run `python stepwavetest.py` when it gave me the error

>   File "stepwavetest.py", line 246
      yield self.steppinmask, 0, 0
     SyntaxError: 'return' with argument inside generator

I'm not really sure what to do now because I looked online and it told me to add a return after every yield but I don't think that is right.
Thanks,

**Comment by pootle** [comment]
That's bizarre - you are running this with python3?

**Comment by GForceSquared** [comment]
I'm not sure, how can I check if I am? Thanks for the help.

**Comment by pootle** [comment]
how are you running the software? 

Something like 
> python3
> import stepwavetest.py  ?

Not 
> python
> import stepwavetest.py`
    - this defaults to python v2

You'll get something like:
> Python 3.5.3 (default, Jan 19 2017, 14:11:04) 
> [GCC 6.3.0 20170124] on linux
> Type "help", "copyright", "credits" or "license" for more information.


The 3.5.3 bit on the first line is the important part - first 3 means Python 3

**Comment by GForceSquared** [comment]
Ok I did

>

>

**Comment by GForceSquared** [comment]
Ok so what I did was ` python3` in my home directory which I then got what you said before. Then I did `import stepwavetest.py` and I got this error. 

> Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named 'stepwavetest'

To be honest, I'm actually unsure how to run your program so that it can drive my stepper motors as I'm relatively new to all of this command line stuff. If you could help me out that would be great.

**Comment by pootle** [comment]
my mistake, it should be 

> import stepwavetest
NOT

> import stepwavetest.py

Have you got pigpio installed and running?

**Comment by GForceSquared** [comment]
Thanks for the help! I have pigpio running and I did `import stepwavetest` and at least to me nothing happened.

>>> import stepwavetest
>>>

 I'm not really sure what to run so that the stepper motors can run with your code. I'm sorry if I'm being annoying to you, I really am over my head with this kind of stuff. Thanks again for your time!

**Comment by GForceSquared** [comment]
For pigpio I did `sudo pigpiod`

**Comment by pootle** [comment]
That's OK, good to help you get started.
You've not said what kit you actually got and how it's all hooked up - that would be useful so I don't need ask a lot of very basic questions.

You need to check that the pin numbers as in line 602/3 are right, and if everything else is setup the runmotors method with the right parameters should get the motor moving.

the parameters (apart from the motor name) are explained in the pulser method around line 206.
probably reduce the values of startsr and maxsr until you know how fast your motors can go

**Comment by GForceSquared** [comment]
Sorry about that I forgot to tell you how I hooked things up. 
So I have two DRV8825 stepper driver chips running two NEMA 17 stepper motors on a Raspberry Pi 3B. Both of them I used the minimum wiring diagram found on Pololu. 

For one of the drivers I have hooked up the STEP pin with pin 16, and DIR with pin 18. For the other I have the STEP pin with pin 15, and DIR with pin 13. These are all GPIO board, not BCM pins.

I looked at lines 602/3 and realized that I didn't change them at all! However, I was wondering if it was necessary to connect to the ENABLE pin with the board, and also if I could get away with only using full steps and not using step levels. 

Just one more question, after I fix the code so that it works with my pi, do I have to run the code by doing `python3` then typing line by line the example you gave us at the bottom of the code?

Thanks for your help, this has been a great learning experience for me.

**Comment by pootle** [comment]
Enable if not connected should mean the driver chip always delivers current. Only connect it to pi if you want to be able to turn it off.

Without the mode mode pins connected it looks like it will default to full steps. Full steps are fine, but tend to be noisy, and some steppers will go faster / are less likely to miss steps in 1/2 step mode (Both types of motor I've tried are better in 1/2 step mode. The other modes can be quieter at lower speeds.

The code is setup like that so you can tweak the parameters to find the best settings for a motor - fastest speed, best ramp rate etc.

Then you can write your own python file that replaces `makeagents` and then calls runmotors with the parameters to match whatever you want to do.

**Comment by GForceSquared** [comment]
Ok thanks for the help, just one final question, are you using GPIO BOARD or are you using GPIO BCM for the pin numbers.

**Comment by GForceSquared** [comment]
Also if it is BCM, where can I change it to BOARD?

**Comment by pootle** [comment]
Had to  look that up - they're BCM numbers as that is what pigpio ( which is what this uses for low level driving) uses. Just [use the info here](https://pinout.xyz/#) to convert the numbers.

**Comment by pootle** [comment]
btw - if you want to run steppers fast the gpio daemon is best run at high priority.

> sudo pigpiod -c 256

**Comment by GForceSquared** [comment]
OK thanks I have the code up and running and now the motors are running, but it doesn't turn in one direction, instead, it **twitches back and forth** a lot. Also I get this error along with running the code. 

> Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/home/pi/pisteppers/stepwavetest.py", line 333, in agentRun
    super().agentRun()
  File "/home/pi/pisteppers/stepwavetest.py", line 65, in agentRun
    inMsgMeth, inMsgKwargs = self.inq.get(block=True, timeout=nexttimerdue-time.time())
  File "/usr/lib/python3.5/queue.py", line 166, in get
    raise ValueError("'timeout' must be a non-negative number")
ValueError: 'timeout' must be a non-negative number


I'm sorry, I don't know what I am doing wrong.

**Comment by pootle** [comment]
Ah! at least it is trying! That is a good sign. Couple of things to check:

have you got the the right pin for step and direction?

It might your pi has got too much to run and the process running the motors is not getting enough CPU.

How long does has it run for when you get the message?

Which type of raspberry pi are you using? Which version of Raspbian are you using? * and is the lite version of the full version?*

run a second shell and use 
> sudo top
while it is running to see how busy the machine is.

Also you could try running it a lot slower to see if it is loosing sync for some reason.

try with startsr at 100 and maxsr at 200 just to see if it steps properly.

Are you using a power supply with the full voltage? What voltage power supply are you using?

Have you set the current limit potentiometer?

**Comment by GForceSquared** [comment]
Ok so I checked the pins and all seem fine, I also run sudo top and saw that pigpiod is using around 5% of the CPU power which looks fine. 
I'm using a Raspberry Pi 3B with the newest **full** version of Raspbian Stretch.

I tried running it with startsr at 10 and maxsr at 100 and the same problem occurs.
My power supply can produce 30 volts and 500mA and I have set the current limit potentiometer.

Also the second I run the code it shows the stepper control and then the error comes up.

One final problem is that while in my code I want both motors to run, right now only one motor is running and the other doesn't move.

**Comment by pootle** [comment]
I'll do a quick fix to avoid the crash (hopefully in a failsafe way)

I've just run this again on a Pi 2B and it is all working OK

**Comment by pootle** [comment]
OK I've updated stepwavetest.py so it should failsafe and avoid the crash. I suspect other processes are delaying the python code at a bad time!

If you used git to download, just do 
> get fetch origin
or update it - I've replaced line 65 with 5 new lines.

Check both motors before running python, but with power on. They should both be very stiff or impossible to move.

Try running the motors 1 at a time and see if they both behave the same way.

It may be that these motors don't like starting in full step mode, you might have to run them in 1/2 step mode - use a pullup on the mode0 pin.
